addons/product/models/product.py

```diff
@@ -880,3 +880,20 @@ def _check_product_variant(self):
         for supplier in self:
             if supplier.product_id and supplier.product_tmpl_id and supplier.product_id.product_tmpl_id != supplier.product_tmpl_id:
                 raise ValidationError(_('The product variant must be a variant of the product template.'))
+
+    @api.model
+    def create(self, vals):
+        # ensure consistency between variant & template
+        # required for supplierinfo to show in product.seller_ids
+        if vals.get('product_id') and not vals.get('product_tmpl_id'):
+            product = self.env['product.product'].browse(vals['product_id'])
+            vals['product_tmpl_id'] = product.product_tmpl_id.id
+        res = super().create(vals)
+        return res
+
+    def write(self, vals):
+        # if template gets deleted, delete variant as well
+        if vals.get('product_tmpl_id') is False:
+            vals['product_id'] = False
```

Comment:
as an alternative for a more rapid feedback, we could use an onchange as they do eg. for mrp.bom:

https://github.com/odoo/odoo/blob/6428081e385ff08e6d4be47938e8b91abeb4246b/addons/mrp/models/mrp_bom.py#L208-L213

I think this would also fix the issue of the create with variant and without product template. This solution would just prevent to do a wrong thing via the interface but they would still be possible by python (but I think that would be ok, you should just not do something wrong in python)

Reply:
> but they would still be possible by python (but I think that would be ok, you should just not do something wrong in python)

Reason I didn't do an `onchange` is because the client did say that ideally they want to be able to create vendor pricelists based on SKUs by importing CSV files, which does make sense if you need to add a record for every product-vendor combination. I assumed `onchange` doesn't work for CSV imports, but haven't actually tested it yet.

Reply:
onchange doesn't work on import (maybe @api.depends do, I'm not sure. If they did we could have product_tmpl_id being a computed store=True readony=False field updated on product_id change)

but the fact that the product_tmpl_id is not filled by product_id doesn't shock me (since in the interface this makes sense, we can only select a product_id if the product_tmpl_id is set), it might be more something like a feature to change something that works

Reply:
Either way, it's currently allowed to have an empty template field, and doesn't work as expected (fetch data when adding the variant to a PO). Rather than restricting it post hoc or adding a non-stored `seller_ids` field to `product.product`, I figured this is the more sensible option.

I could maybe modify it to the approach used on create/write in `product.template` to keep `type` and `detailed_type` in sync by calling a `_sanitize_vals` method https://github.com/odoo/odoo/blob/54eba7567b57ce8d389600988b07358afde2503c/addons/product/models/product_template.py#L445-L457